### DISCUSSION TITLE
prisma migrate: The provided value for the column is too long for the column's type. Column: bodyEmail

### DISCUSSION THREAD

**Discussion by alan345** [prisma/prisma#4079] [Q&A] [answered]
After running 

```
npx prisma migrate save --experimental
npx prisma migrate up --experimental
```

I have this error: 
```
Error: P2000

The provided value for the column is too long for the column's type. Column: bodyEmail
```

This Column `bodyEmail` is from the table `Campaign`:

```sql

show create table Campaign;
 > show create table Campaign;
 "Campaign","CREATE TABLE `Campaign` (
  `id` char(30) CHARACTER SET utf8 COLLATE utf8_general_ci NOT NULL,
  `userId` char(30) CHARACTER SET utf8 COLLATE utf8_general_ci NOT NULL,
  `bodyEmail` mediumtext,
  `from` mediumtext,
  `subject` mediumtext,
  `name` mediumtext,
  `createdAt` datetime(3) NOT NULL,
  `lastSent` datetime(3) DEFAULT NULL,
  `updatedAt` datetime(3) NOT NULL,
  `description` mediumtext NOT NULL,
  `bcc` mediumtext NOT NULL,
  `isActive` tinyint(1) NOT NULL,
  PRIMARY KEY (`id`),
  KEY `userId` (`userId`),
  CONSTRAINT `Campaign_ibfk_1` FOREIGN KEY (`userId`) REFERENCES `User` (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci"
```

**Comment by janpio** [comment]
When do you get this error? You might want to create an issue at https://github.com/prisma/migrate/issues

**Reply by ryands17** [reply]
@alan345 👋 
Do add the details as I cannot reproduce this by just creating your _Campaign_ table in SQL and then running Migrate over it.

**Reply by thrant** [reply]
I'm running into this as well. The base issue seems to be that the String scalar in prisma.schema maps to varchar(191) in MySQL (instead of "text"), despite being mapped to "text" in PostgreSQL and SqlLite. 

This field is obviously too short for many usecases, so we run the migration but need to manually change the column in the DB to be a longer varchar, or mediumtext, or text, or anything like that. All good so far, at least in greenfield development.

Now then DB gets populated with a row where the length in that String column exceeds 191 chars, all good so far.

However now we change something in the shema.prisma, even something totally unrelated to that text in question, and try to run prisma:save && prisma:up -> it fails with the "too long" error, since it finds that somewhere in the DB a field exists which is mapped as String but the contents exceeds 191 chars. 

The fixes could be to.. 
A) Separate String and LongString or something like that
B) Change the DB column type to be "text" in MySQL like it's in the others
C) Let the Prisma migrate run through even if it detects something too long, as long as it does not need to modify anything in that column

**Reply by alan345** [reply]
Regarding my example, here you have a few data: 
```
mysql> SELECT length(bodyEmail) as lengthBodyEmail, id  FROM Campaign order By lengthBodyEmail desc limit 10;
+-----------------+-----------------------------+
| lengthBodyEmail | id                          |
+-----------------+-----------------------------+
|            1274 | ckevv25gg09885ulv6nu57p3k   |
|            1077 | ckh0p3cdi556104aklvu2axqcqy |
|            1077 | ckh0s9hjy573020aklv4gzsbep0 |
|            1075 | ckh0rvbza569370aklvwq4h1d86 |
|            1075 | ckh0monu6547886aklvmn4mbad7 |
|            1022 | ckh0p823z556695aklvamxu98kk |
|            1022 | ckh0sfcjs578385aklvenfpah2q |
|            1002 | ckfcranua279876jzlv1g1492kl |
|             998 | ckg6ycnac6364hilvbqdvi2vo   |
|             994 | ckg6y9xcf4597hilv2ojs3ig4   |
+-----------------+-----------------------------+
10 rows in set (0.00 sec)
```
Most of bodyEmail has more than 191 chars. 

Why deciding to set a max of 191 chars (`varchar(191)`)? https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference/#string ?
Thanks

**Reply by ryands17** [reply, marked as answer]
@alan345 Yes that was something added at the beginning when this was designed. Currently you would can directly change the value from the DB, after running Migrate commands.

If you're in the Slack channel, you must have seen the the Early Access Migrate [doc](https://www.notion.so/prismaio/Documentation-for-Prisma-Migrate-Early-Access-7708df0c1d3d4852ab39fb69ed182786).
With this, you can directly set the datatype of the fields via SQL.

**Comment by thrant** [comment]
FYI, looks like there's also an Issue open for this (since 2019): https://github.com/prisma/migrate/issues/116

**Comment by larinius** [comment]
What purpose to use automatic migrations if we must fix database manually every time just because Prisma decided that191 character is enough for every project? 
I have text description for product category, and cant apply it, just because of that error.
At least put a warning on Prisma description - "It cannot handle more than 191 characters", so people will pick another more capable ORM.

**Reply by martyaxel** [reply]
I believe you can, starting from prisma 2.15.0.
Example: 
model Test{
  title   String @db.VarChar(1000)
  description    String         @db.Text
}

I have just used @db.Text and it worked for me.
https://www.prisma.io/docs/concepts/components/prisma-schema/data-model#native-types-mapping

**Reply by larinius** [reply]
Yes, after finding same issues on SO, I use "db.Text" + "db.varchar(N)" on every field. 
There is no information about this "feature" in Prisma documentation.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
